presto-main/src/main/java/com/facebook/presto/sql/planner/OptimizerStatsRecorder.java

```diff
@@ -0,0 +1,79 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.sql.planner;
+
+import com.facebook.presto.sql.planner.optimizations.OptimizerStats;
+import com.facebook.presto.sql.planner.optimizations.PlanOptimizer;
+import org.weakref.jmx.MBeanExporter;
+import org.weakref.jmx.ObjectNames;
+
+import java.util.HashMap;
+import java.util.Map;
+
+import static com.google.common.base.Preconditions.checkArgument;
+import static com.google.common.base.Verify.verify;
+import static java.lang.String.format;
+import static java.util.Objects.requireNonNull;
+
+public class OptimizerStatsRecorder
+{
+    private final Map<Class<?>, OptimizerStats> stats = new HashMap<>();
+
+    public void register(PlanOptimizer optimizer)
+    {
+        requireNonNull(optimizer, "optimizer is null");
+        checkArgument(!optimizer.getClass().isAnonymousClass());
+        stats.put(optimizer.getClass(), new OptimizerStats());
+    }
+
+    public void record(PlanOptimizer optimizer, long nanos)
+    {
+        requireNonNull(optimizer, "optimizer is null");
+        stats.get(optimizer.getClass()).record(nanos);
+    }
+
+    public void recordFailure(PlanOptimizer optimizer)
+    {
+        requireNonNull(optimizer, "optimizer is null");
+        stats.get(optimizer.getClass()).recordFailure();
+    }
+
+    void export(MBeanExporter exporter)
+    {
+        for (Map.Entry<Class<?>, OptimizerStats> entry : stats.entrySet()) {
+            verify(!entry.getKey().getSimpleName().isEmpty());
+            String name = ObjectNames.builder(PlanOptimizer.class)
+                    .withProperty("optimizer", entry.getKey().getSimpleName())
+                    .build();
+
+            try {
+                exporter.export(name, entry.getValue());
+            }
+            catch (RuntimeException e) {
+                throw new RuntimeException(format("Failed to export MBean with name '%s'", name), e);
+            }
+        }
+    }
+
+    void unexport(MBeanExporter exporter)
+    {
+        for (Class<?> rule : stats.keySet()) {
+            String name = ObjectNames.builder(PlanOptimizer.class)
+                    .withProperty("optimizer", rule.getSimpleName())
+                    .build();
```

Comment:
Consider extracting object name building to a method to emphasize that export and unexport use the same naming strategy.